_posts/2019-10-17-batchowe-inserty-w-hibernate-droga-ku-szybkosci.md

```diff
@@ -0,0 +1,139 @@
+ 
+W praktyce oznacza to, że mogą powstać „dziury” w numeracji, ponieważ jeżeli w danej transakcji wstawimy jedną encję to mimo wszystko potrzebujemy numeru sekwencji z bazy danych, a w konsekwencji podbijemy ją o N, ale zyskiem korzystania z tego mechanizmu jest rzadka potrzeba pytania bazy o kolejną wartość sekwencji.
+Sprawdźmy więc kolejny raz jak nasze optymalizacje wpłynęły na szybkość działania usługi.
+
+![Zipkin - przepływ na małej ilości encji po drugiej optymalizacji](/assets/img/posts/2019-10-17-batchowe-inserty-w-hibernate-droga-ku-szybkosci/grafika3.png)
+
+Dla dużej ilości encji statystyki prezentują się następująco:
+```
+Łączny czas odpowiedzi usługi przy wywołaniu przez HTTP: 1746 ms (POSTMAN)
+Kluczowe statystyki wygenerowane przez Hibernate:
+4294453 nanoseconds spent preparing 26 JDBC statements;
+118572140 nanoseconds spent executing 23 JDBC statements;
+531069793 nanoseconds spent executing 4 JDBC batches;
+732899796 nanoseconds spent executing 1 flushes (flushing a total of 2003 entities and 0 collections);
+5497 nanoseconds spent executing 1 partial-flushes (flushing a total of 0 entities and 0 collections)
+```
+Widzimy kolejny drastyczny spadek ilości wykonywanych operacji wynikający z tego, że nie musimy tak często odpytywać bazy o kolejne numery sekwencji.
+Porównajmy teraz łączne wywołania usługi
+```
+Stan początkowy – bez optymalizacji: 9503 ms
+Z batchowymi insertami: 5138 ms
+Z batchowymi insertami i optymalizacją sekwencji: 1746 ms
+```
+Dzięki użyciu odpowiednich mechanizmów zmniejszyliśmy czas odpowiedzi naszej usługi o ponad 80%!
+Warto wspomnieć, że przy tego typu optymalizacjach zmiany w czasach wykonywania żądań mogą być wysoce zależne od użytych technologii, dla konkretnych baz danych istnieją też specyficzne możliwe optymalizacje. Dobrym przykładem tego jest ustawienie przydatne jeśli używamy MySQL
+```properties
+rewriteBatchedStatements=true
+```
+pozwalające na zoptymalizowanie ilości zapytań przesyłanych w pakiecie sieciowym.
+ 
+W przypadku optymalizacji takich jak pokazane w tym poście warto zajrzeć pod maskę - do dokumentacji, a nawet kodu źródłowego frameworku, a także zaopatrzyć się w dobre narzędzia wspomagające diagnostykę jak chociażby wyżej przytoczony Zipkin.
+
+Warto też strzec się błędów i niepoprawnych użyć mechanizmów frameworków, należy chociażby pamiętać o tym, że aby wykonać batch insert za pomocą klasy JpaRepository z frameworku Spring musimy skorzystać z metody *saveAll*, a nie *save*. Osobnej optymalizacji poprzez ustawienie *order_updates* w plikach konfiguracyjnych może wymagać również uaktualnianie rekordów. Kolejnym problemem o którym należy pamiętać jest fakt, że cache poziomu L1 (tożsame z PersistanceContext) będące cache poziomu transakcji jest zawsze włączone, co oznacza, że zapisując dużą ilość encji w jednej transakcji narażamy się na problemy pamięciowe.
```

Comment:
To zdanie z cacheami trzeba przeredagować, bo jest niezrozumiałe. Coś tam z odmianą i liczbą się pokręciło